core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -0,0 +1,157 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.coordinator.transaction
+
+import java.util
+
+import kafka.common.{InterBrokerSendThread, RequestAndCompletionHandler}
+import kafka.server.{DelayedOperationPurgatory, KafkaConfig, MetadataCache}
+import kafka.utils.Logging
+import org.apache.kafka.clients._
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.network._
+import org.apache.kafka.common.requests.{TransactionResult, WriteTxnMarkersRequest}
+import org.apache.kafka.common.security.JaasContext
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.common.Node
+
+import scala.collection.mutable
+import java.util.concurrent.BlockingQueue
+
+import collection.JavaConverters._
+import collection.JavaConversions._
+
+case class CoordinatorEpochAndMarkers(coordinatorEpoch: Int, txnMarkerEntry: util.List[WriteTxnMarkersRequest.TxnMarkerEntry])
+case class DestinationBrokerAndQueuedMarkers(destBrokerNode: Node, markersQueue: BlockingQueue[CoordinatorEpochAndMarkers])
+
+object TransactionMarkerChannelManager {
+  def apply(config: KafkaConfig,
+            metrics: Metrics,
+            metadataCache: MetadataCache,
+            txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
+            time: Time): TransactionMarkerChannelManager = {
+
+    val channel = new TransactionMarkerChannel(config.interBrokerListenerName, metadataCache)
+    val sendThread: InterBrokerSendThread = {
+      val threadName = "TxnMarkerSenderThread-" + config.brokerId
+      val channelBuilder = ChannelBuilders.clientChannelBuilder(
+        config.interBrokerSecurityProtocol,
+        JaasContext.Type.SERVER,
+        config,
+        config.interBrokerListenerName,
+        config.saslMechanismInterBrokerProtocol,
+        config.saslInterBrokerHandshakeRequestEnable
+      )
+      val selector = new Selector(
+        NetworkReceive.UNLIMITED,
+        config.connectionsMaxIdleMs,
+        metrics,
+        time,
+        "replica-fetcher",
+        Map("broker-id" -> config.brokerId.toString).asJava,
+        false,
+        channelBuilder
+      )
+      val networkClient = new NetworkClient(
+        selector,
+        new ManualMetadataUpdater(),
+        threadName,
+        1,
+        0,
```

Comment:
I think @guozhangwang would have the most context on that, but I don't think we should set it to 0. This is the connection to the leaders to write the abort marker. If the connection is disconnected, there is little reason to try to establish it again. I think the default of 50ms is reasonable here.